YeelightAPI/DeviceLocator.cs

```diff
@@ -14,244 +16,441 @@
-        private static readonly char[] _colon = new char[] { ':' };
-        private static readonly IPEndPoint _multicastEndPoint = new IPEndPoint(IPAddress.Parse("239.255.255.250"), 1982);
-        private static readonly byte[] _ssdpDiagram = Encoding.ASCII.GetBytes(_ssdpMessage);
-        private static string _yeelightlocationMatch = "Location: yeelight://";
-
-        #endregion Private Fields
-
-        /// <summary>
-        /// Notification Received event
-        /// </summary>
-        public static event DeviceFoundEventHandler OnDeviceFound;
-
-        /// <summary>
-        /// Notification Received event handler
-        /// </summary>
-        /// <param name="sender"></param>
-        /// <param name="e"></param>
-        public delegate void DeviceFoundEventHandler(object sender, DeviceFoundEventArgs e);
-
-        #region Public Methods
-
-        /// <summary>
-        /// Discover devices in a specific Network Interface
-        /// </summary>
-        /// <param name="preferedInterface"></param>
-        /// <returns></returns>
-        public static async Task<List<Device>> Discover(NetworkInterface preferedInterface)
-        {
-            List<Task<List<Device>>> tasks = CreateDiscoverTasks(preferedInterface);
-            List<Device> devices = new List<Device>();
+    [Obsolete("Deprecated. Event will be removed in next release.")]
+    public static event DeviceFoundEventHandler OnDeviceFound;
 
-            if (tasks.Count != 0)
-            {
-                await Task.WhenAll(tasks);
+    /// <summary>
+    ///   Notification Received event handler
+    /// </summary>
+    /// <param name="sender"></param>
+    /// <param name="e"></param>
+    [Obsolete("Deprecated. Event will be removed in next release.")]
+    public delegate void DeviceFoundEventHandler(object sender, DeviceFoundEventArgs e);
 
-                devices.AddRange(tasks.SelectMany(t => t.Result).GroupBy(d => d.Hostname).Select(g => g.First()));
-            }
+    #region Public Methods
 
-            return devices;
-        }
+    /// <summary>
+    ///   Discover devices in a specific Network Interface
+    /// </summary>
+    /// <param name="preferedInterface"></param>
+    /// <returns></returns>
+    [Obsolete("Deprecated. Use DiscoverAsync and overloads instead.")]
+    public static async Task<List<Device>> Discover(NetworkInterface preferedInterface)
+    {
+      List<Task<List<Device>>> tasks = DeviceLocator.CreateDiscoverTasks(preferedInterface);
 
-        /// <summary>
-        /// Discover devices in LAN
-        /// </summary>
-        /// <returns></returns>
-        public static async Task<List<Device>> Discover()
-        {
-            List<Task<List<Device>>> tasks = new List<Task<List<Device>>>();
-            List<Device> devices = new List<Device>();
+      List<Device>[] result = await Task.WhenAll(tasks);
+      return result.SelectMany(devices => devices).ToList();
```

Comment:
Even if this method is now obsolete, we need to keep the previous code to ensure unicity of found devices : 
`return result.SelectMany(devices => devices).`**GroupBy(d => d.Hostname).Select(g => g.First())**`.ToList()`

Reply:
Sorry. Fixed. I've also added the grouping to the new `DiscoverAsync`.  

I also refactored the `CheckSocketForDevices` method (extracted to new method).